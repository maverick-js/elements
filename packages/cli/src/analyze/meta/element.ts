import type ts from 'typescript';

import type { DocTagMeta, FileMeta, TS_NODE } from './component';

export interface TagMeta {
  [TS_NODE]: ts.PropertyDeclaration;
  name: string;
}

export interface AttrMeta {
  [TS_NODE]: ts.Node;
  attr: string | false;
}

export interface AttrsMeta {
  [propName: string]: AttrMeta;
}

export interface ElementMeta {
  [TS_NODE]: ts.ClassDeclaration;
  name: string;
  file: FileMeta;
  tag: TagMeta;
  component: {
    [TS_NODE]: ts.ClassDeclaration;
    name: string;
  };
  docs?: string;
  doctags?: DocTagMeta[];
  attrs?: AttrsMeta;
}
