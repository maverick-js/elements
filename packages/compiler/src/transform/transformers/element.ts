import {
  $,
  addClassMembers,
  createStaticComputedProperty,
  createSymbolFor,
  findImportDeclarationFromModule,
  findImportSpecifierFromDeclaration,
  findStaticProp,
  isClassThatExtends,
} from '@maverick-js/ts';
import ts from 'typescript';

export function markCustomElements(sourceFile: ts.SourceFile) {
  const maverickImport = findImportDeclarationFromModule(sourceFile, 'maverick.js'),
    maverickComponent = findImportSpecifierFromDeclaration(maverickImport, 'Component');

  if (maverickComponent) {
    const statements: ts.Statement[] = [];

    for (const statement of sourceFile.statements) {
      if (isMaverickElementClass(statement, maverickComponent)) {
        const newClass = addClassMembers(statement, [createStaticElementProp()]);
        statements.push(newClass);
      } else {
        statements.push(statement);
      }
    }

    return $.updateSourceFile(sourceFile, statements);
  }

  return sourceFile;
}

function createStaticElementProp() {
  return createStaticComputedProperty(createSymbolFor('element'), $.createTrue());
}

function isMaverickElementClass(
  node: ts.Node,
  component: ts.ImportSpecifier,
): node is ts.ClassDeclaration {
  return isClassThatExtends(node, component.name) && !!findStaticProp(node, 'tagName');
}
