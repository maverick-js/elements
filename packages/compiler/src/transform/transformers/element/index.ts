import ts from 'typescript';

import type { Transformer } from '../transformer';

export const ElementTransformer: Transformer = {
  name: '@maverick-js/element',
  transform({ code, sourceFile, jsx, ctx }) {},
};

// Example
// class Foo extends Component {
//   static tagName = 'mk-foo';
//   override render(): JSX.Element {
//     return <host as="div"></host>;
//   }
// }

// class FooElement extends Host(HTMLElement, Foo) {
//   static tagName = 'mk-foo';
//   override onConnect(el: HTMLElement) {
//     // rendering here??
//   }
// }
