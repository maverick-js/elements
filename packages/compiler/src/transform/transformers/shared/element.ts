import {
  $,
  addClassMembers,
  createImports,
  createStaticComputedMethod,
  findImportDeclarationFromModule,
  findImportSpecifierFromDeclaration,
  findStaticProp,
  isClassThatExtends,
} from '@maverick-js/ts';
import ts from 'typescript';

export function setupCustomElements(
  sourceFile: ts.SourceFile,
  register: (component: ts.ClassDeclaration) => ts.Expression,
  registerImports: ts.Statement[] = [],
): ts.SourceFile {
  const maverickImport = findImportDeclarationFromModule(sourceFile, 'maverick.js'),
    maverickComponent = findImportSpecifierFromDeclaration(maverickImport, 'Component');

  if (maverickComponent) {
    let statements: ts.Statement[] = [],
      hasCustomElement = false;

    for (const statement of sourceFile.statements) {
      if (isMaverickElementClass(statement, maverickComponent)) {
        const newClass = addClassMembers(statement, [
          createStaticElementRegistration(register(statement)),
        ]);

        statements.push(newClass);
        hasCustomElement = true;
      } else {
        statements.push(statement);
      }
    }

    return hasCustomElement
      ? $.updateSourceFile(sourceFile, [
          ...registerImports,
          createCustomElementSymbolImport(),
          ...statements,
        ])
      : sourceFile;
  }

  return sourceFile;
}

const CUSTOM_ELEMENT_SYMBOL_NAME = 'CUSTOM_ELEMENT_SYMBOL';

function createCustomElementSymbolImport() {
  return createImports([$.id(CUSTOM_ELEMENT_SYMBOL_NAME)], 'maverick.js');
}

function createStaticElementRegistration(registration: ts.Expression) {
  return createStaticComputedMethod(
    $.id(CUSTOM_ELEMENT_SYMBOL_NAME),
    [],
    [$.createReturnStatement(registration)],
  );
}

function isMaverickElementClass(
  node: ts.Node,
  specifier: ts.ImportSpecifier,
): node is ts.ClassDeclaration {
  return isClassThatExtends(node, specifier.name) && !!findStaticProp(node, 'element');
}
