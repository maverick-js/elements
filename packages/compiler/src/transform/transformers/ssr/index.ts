import ts from 'typescript';

import type { Transformer } from '../transformer';

export const SSRTransformer: Transformer = {
  name: '@maverick-js/ssr',
  transform({ code, sourceFile, jsx, ctx }) {},
};
