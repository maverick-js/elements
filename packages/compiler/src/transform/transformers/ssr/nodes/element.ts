import { escapeHTML } from '@maverick-js/std';
import type ts from 'typescript';

import { type AttributeNode, type ElementNode, isFragmentNode } from '../../../../parse/ast';
import { trimQuotes } from '../../../../utils/print';
import { $, createElementProps } from '../../ts-factory';
import type { SsrVisitorContext } from '../state';

export function Element(node: ElementNode, { state, walk }: SsrVisitorContext) {
  const { runtime } = state,
    isRoot = walk.path.length === 0 || isFragmentNode(walk.path.at(-1)!);

  if (node.isDynamic()) {
    state.marker();
  }

  state.html += `<${node.as ?? node.name}`;

  if (node.spreads) {
    const props = runtime.mergeProps([
      ...node.spreads.map((s) => s.initializer),
      createElementProps(node, { ssr: true }),
    ]);

    state.value(runtime.attrs(props));
  } else {
    let classAttr: AttributeNode | null = null,
      stylesAttr: AttributeNode | null = null,
      props: ts.PropertyAssignment[] = [];

    function addProp(attr: AttributeNode, dynamic = attr.dynamic) {
      if (dynamic) {
        props.push($.createPropertyAssignment($.string(attr.name), attr.initializer));
      } else {
        state.html += ` ${attr.name}="${escapeHTML(trimQuotes(attr.initializer.getText()), true)}"`;
      }
    }

    if (node.attrs) {
      for (const attr of node.attrs) {
        if (attr.name === 'class') {
          classAttr = attr;
        } else if (attr.name === 'style') {
          stylesAttr = attr;
        } else {
          addProp(attr);
        }
      }
    }

    if (node.classes) {
      const base = classAttr?.initializer ?? $.string(''),
        $classProps = node.classes?.map((c) =>
          $.createPropertyAssignment($.string(c.name), c.initializer),
        ),
        classProps = $classProps ? runtime.class(base, $classProps) : base;

      props.push($.createPropertyAssignment('class', classProps));
    } else if (classAttr) {
      // Handle roots at runtime because of host props on component <Foo class={} $class:foo={}>.
      addProp(classAttr, isRoot || classAttr.dynamic);
    }

    if (node.styles || node.vars) {
      const base = stylesAttr?.initializer ?? $.string(''),
        $styleProps = node.styles?.map((s) =>
          $.createPropertyAssignment($.string(s.name), s.initializer),
        ),
        $varProps = node.vars?.map((s) =>
          $.createPropertyAssignment($.string(`--${s.name}`), s.initializer),
        ),
        styleProps =
          $styleProps || $varProps
            ? runtime.style(base, [...($styleProps ?? []), ...($varProps ?? [])])
            : base;

      props.push($.createPropertyAssignment('style', styleProps));
    } else if (stylesAttr) {
      // Handle roots at runtime because of host props on component <Foo $var:foo={}>.
      addProp(stylesAttr, isRoot || stylesAttr.dynamic);
    }

    if (props.length > 0) {
      state.value(runtime.attrs($.createObjectLiteralExpression(props, true)));
    }
  }

  if (node.isVoid) {
    state.html += '/>';
  } else {
    state.html += '>';

    if (node.content) {
      state.value(node.content.initializer);
    } else {
      walk.children();
    }

    state.html += `</${node.as ?? node.name}>`;
  }
}
