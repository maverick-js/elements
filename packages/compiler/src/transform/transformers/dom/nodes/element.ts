import { trimQuotes } from '@maverick-js/std';
import { $ } from '@maverick-js/ts';
import type ts from 'typescript';

import {
  type AstNode,
  type AttributeNode,
  type ElementNode,
  isElementNode,
  isFragmentNode,
} from '../../../../parse/ast';
import { getAttributeText } from '../../../../parse/utils';
import { createElementSpreadProps } from '../../shared/factory';
import { getElementId } from '../position';
import type { DomRuntime } from '../runtime';
import type { DomVisitorContext } from '../state';

export function Element(node: ElementNode, { state, walk }: DomVisitorContext) {
  const { hydratable, vars, runtime, block } = state,
    isRoot = walk.path.length === 0 || isFragmentNode(walk.path.at(-1)!);

  if (!state.template) {
    state.template = $.createUniqueName('$_template');
    state.importNodes = containsCustomElement(node);
  }

  if (state.hydratable && (isRoot || node.isDynamic())) {
    state.html += '<!$>';
  }

  if (isRoot) {
    if (hydratable) {
      const bindings = vars.setup.walker(state.template);
      state.element = bindings.root;
      state.walker = bindings.walker;
    } else {
      state.element = vars.setup.root(state.template);
    }
  } else if (node.isDynamic()) {
    if (state.hydratable) {
      state.element = vars.setup.nextElement(state.walker!);
    } else {
      state.element = getElementId(node, state, walk);
    }
  }

  if (isRoot && state.element) {
    state.elements.set(node, state.element);
  }

  state.html += `<${node.name}`;

  if (state.element) {
    if (node.spreads) {
      const props = runtime.mergeProps([
        ...node.spreads.map((s) => s.initializer),
        createElementSpreadProps(node),
      ]);

      block.push(runtime.spread(state.element, props));
    } else {
      block.push(
        ...createElementDomExpressions(state.element, node, runtime, {
          delegatedEvents: state.delegatedEvents,
          onStaticAttr: (attr) => {
            state.html += ` ${attr.name}="${getAttributeText(attr)}"`;
          },
        }),
      );
    }
  }

  if (state.element && state.result === $.null) {
    state.result = state.element;
  }

  if (node.isVoid) {
    state.html += '/>';
  } else {
    state.html += '>';
    if (node.content) {
      if (node.content.dynamic) {
        if (state.element) {
          block.push(runtime.content(state.element, node.content.name, node.content.initializer));
        }
      } else {
        state.html += trimQuotes(node.content.initializer.getText());
      }
    } else {
      walk.children();
    }

    state.html += `</${node.name}>`;
  }
}

export function containsCustomElement(node: AstNode) {
  if (isElementNode(node) && node.isCustomElement) {
    return true;
  }

  if ((isElementNode(node) || isFragmentNode(node)) && node.children) {
    for (const child of node.children) {
      if (containsCustomElement(child)) return true;
    }
  }

  return false;
}

export function createElementDomExpressions(
  el: ts.Identifier,
  node: ElementNode,
  runtime: DomRuntime,
  {
    delegatedEvents,
    onStaticAttr,
  }: {
    delegatedEvents?: Set<string>;
    onStaticAttr: (attr: AttributeNode) => void;
  },
) {
  const block: ts.Expression[] = [];

  if (node.attrs) {
    for (const attr of node.attrs) {
      if (!attr.dynamic) {
        onStaticAttr(attr);
      } else if (attr.signal && attr.name === 'class' && node.classes) {
        block.push(runtime.classTokens(el, attr.initializer));
      } else if (attr.signal && attr.name === 'style' && (node.styles || node.vars)) {
        block.push(runtime.styleTokens(el, attr.initializer));
      } else {
        block.push(runtime.attr(el, attr.name, attr.initializer));
      }
    }
  }

  if (node.props) {
    for (const prop of node.props) {
      block.push(runtime.prop(el, prop.name, prop.initializer, prop.signal));
    }
  }

  if (node.classes) {
    for (const c of node.classes) {
      block.push(runtime.class(el, c.name, c.initializer));
    }
  }

  if (node.styles) {
    for (const style of node.styles) {
      block.push(runtime.style(el, style.name, style.initializer));
    }
  }

  if (node.vars) {
    for (const cssvar of node.vars) {
      block.push(runtime.style(el, `--${cssvar.name}`, cssvar.initializer));
    }
  }

  if (node.events) {
    for (const event of node.events) {
      block.push(runtime.listen(el, event.type, event.initializer, event.capture));
      if (event.delegate) delegatedEvents?.add(event.type);
    }
  }

  if (node.ref) {
    block.push(runtime.ref(el, node.ref.initializer));
  }

  return block;
}
