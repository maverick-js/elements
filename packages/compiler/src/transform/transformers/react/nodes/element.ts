import { trimQuotes } from '@maverick-js/std';
import { $, isAccessExpression } from '@maverick-js/ts';
import type ts from 'typescript';

import {
  type AstNode,
  type AttributeNode,
  type ElementNode,
  type FragmentNode,
  isElementNode,
  isFragmentNode,
  isTextNode,
} from '../../../../parse/ast';
import { getElementDepth, isStaticTree } from '../../../../parse/utils';
import { containsCustomElement, createElementDomExpressions } from '../../dom/nodes/element';
import { createElementSpreadProps } from '../../shared/factory';
import { getReactPropName } from '../attr-map';
import { createDangerouslySetInnerHTMLProp, createReactNode } from '../react-node';
import type { ReactTransformState, ReactVisitorContext } from '../state';

export function Element(node: ElementNode, { state, walk }: ReactVisitorContext) {
  const { runtime, domRuntime } = state;

  if (isStaticTree(node)) {
    if (node.children && getElementDepth(node) >= 20) {
      // If the tree is more than 20 elements deep we can improve performance by cloning and
      // inserting the HTML.
      renderTemplate(node, state);
    } else {
      // Hoist static virtual nodes.
      hoistRender(node, state);
    }
  } else {
    const mode = state.isExpressionChild ? 'render' : 'setup';

    let el = $.id('el'),
      props: ts.ObjectLiteralElementLike[] = [],
      attach: ts.Expression[] = [];

    if (node.spreads) {
      const mergedProps = domRuntime.mergeProps([
        ...node.spreads.map((s) => s.initializer),
        createElementSpreadProps(node),
      ]);

      const propsId = state[mode].vars.create(
        '$_spread_props',
        mode === 'render'
          ? runtime.memo(
              state.currentScope,
              mergedProps,
              node.spreads.map((s) => s.initializer).filter(isAccessExpression),
            )
          : mergedProps,
      ).name;

      const ssrProps = runtime.spread(propsId),
        ssrPropsId = state[mode].vars.create(
          '$_ssr_props',
          runtime.ifServer(
            mode === 'render' ? runtime.memo(state.currentScope, ssrProps, [propsId]) : ssrProps,
            $.null,
          ),
        ).name;

      attach.push(domRuntime.spread(el, propsId));

      props.push($.createSpreadAssignment(ssrPropsId));
    } else {
      const clientProps = getClientProps(el, node, state);
      props.push(...clientProps.props);
      attach.push(...clientProps.attach);

      const ssrProps = getSsrProps(node, state);

      if (ssrProps.length > 0) {
        const ssrAttrs = $.object(ssrProps, true),
          ssrAttrsId = state[mode].vars.create(
            '$_ssr_attrs',
            runtime.ifServer(
              mode === 'render' ? runtime.memo(state.currentScope, ssrAttrs) : ssrAttrs,
              $.null,
            ),
          ).name;

        props.push($.createSpreadAssignment(ssrAttrsId));
        props.push($.createPropertyAssignment('suppressHydrationWarning', $.createTrue()));
      }
    }

    if (node.content) {
      if (node.content.signal) {
        attach.push(domRuntime.content(el, node.content.name, node.content.initializer));
      } else {
        props.push(createDangerouslySetInnerHTMLProp(node.content.initializer));
      }
    }

    if (attach.length > 0) {
      const attachId = $.createUniqueName('$_attach'),
        attachCallback = $.fn(attachId, [el], attach);

      if (!node.isDynamic()) {
        props.push($.createPropertyAssignment('ref', attachId));
      } else {
        const refId = state.setup.vars.create('$_ref', runtime.signal($.null)).name;
        state.setup.block.push(runtime.onAttach(refId, attachId));
        props.push($.createPropertyAssignment('ref', $.prop(refId, 'set')));
      }

      const scope = node.isDynamic() ? 'setup' : 'module';
      state[scope].block.push(attachCallback);
    }

    // Hoist static props.
    const shouldHoistProps = !node.isDynamic() && props.length > 0,
      propsId = shouldHoistProps
        ? state.module.vars.create('$_props', $.pure($.object(props, true))).name
        : null;

    const vNode = createReactNode(node.name, propsId ?? props);
    state.appendNode(vNode, !node.content ? walk : undefined);
  }
}

export function hoistRender(node: ElementNode | FragmentNode, state: ReactTransformState) {
  const id = $.createUniqueName('$_static_node');
  state.module.vars.create(id, $.pure(state.runtime.createStaticElement(node)));
  if (!state.result) state.result = id;
  if (state.node) state.node.children.push(id);
}

export function renderTemplate(node: ElementNode, state: ReactTransformState) {
  const { runtime, domRuntime } = state;

  const vNodeId = $.createUniqueName('$_node'),
    templateId = $.createUniqueName('$_template'),
    refCallback = runtime.appendHtml($.call(templateId)),
    refProp = $.createPropertyAssignment('ref', refCallback),
    vNode = $.pure(runtime.createElement(createReactNode(node.name, [refProp]))),
    template = $.string(node.children!.map(renderToString).join(''));

  state.module.vars.create(vNodeId, vNode);
  state.module.vars.create(
    templateId,
    domRuntime.createTemplate(template, containsCustomElement(node)),
  );

  if (!state.result) state.result = vNodeId;
}

export function renderToString(node: AstNode) {
  if (isElementNode(node)) {
    let html = `<${node.name}>`;

    if (node.content) {
      html += trimQuotes(node.content.initializer.getText());
    } else if (node.children) {
      for (const child of node.children) {
        html += renderToString(child);
      }
    }

    return html + `</${node.name}>`;
  } else if (isFragmentNode(node)) {
    if (node.children) {
      let html = '';

      for (const child of node.children) {
        html += renderToString(child);
      }

      return html;
    }
  } else if (isTextNode(node)) {
    return node.value;
  } else {
    return '';
  }
}

function getClientProps(
  el: ts.Identifier,
  node: ElementNode,
  { domRuntime, delegatedEvents }: ReactTransformState,
) {
  const props: ts.PropertyAssignment[] = [],
    attach: ts.Expression[] = [
      ...createElementDomExpressions(el, node, domRuntime, {
        delegatedEvents,
        onStaticAttr: (attr) => {
          const name = getReactPropName(attr.name);
          props.push($.createPropertyAssignment(name, attr.initializer));
        },
      }),
    ];

  return { props, attach };
}

export function getSsrProps(node: ElementNode, { runtime, ssrRuntime }: ReactTransformState) {
  let classNode: AttributeNode | null = null,
    stylesNode: AttributeNode | null = null,
    props: ts.PropertyAssignment[] = [];

  function addProp(attr: AttributeNode) {
    const name = getReactPropName(attr.name);
    props.push(
      $.createPropertyAssignment(
        $.string(name),
        attr.signal ? ssrRuntime.unwrap(attr.initializer) : attr.initializer,
      ),
    );
  }

  if (node.attrs) {
    for (const attr of node.attrs) {
      if (attr.name === 'class') {
        classNode = attr;
      } else if (attr.name === 'style') {
        stylesNode = attr;
      } else if (attr.dynamic) {
        addProp(attr);
      }
    }
  }

  if (node.classes) {
    const base =
        classNode?.signal && classNode.initializer
          ? ssrRuntime.unwrap(classNode.initializer)
          : (classNode?.initializer ?? $.string('')),
      $classProps = node.classes?.map((c) =>
        $.createPropertyAssignment($.string(c.name), c.initializer),
      ),
      classProps = $classProps ? ssrRuntime.class(base, $classProps) : base;

    props.push($.createPropertyAssignment('className', classProps));
  } else if (classNode) {
    const value = classNode.signal
      ? ssrRuntime.unwrap(classNode.initializer)
      : classNode.initializer;
    props.push($.createPropertyAssignment('className', ssrRuntime.class(value, [])));
  }

  if (node.styles || node.vars) {
    const base =
        stylesNode?.signal && stylesNode.initializer
          ? ssrRuntime.unwrap(stylesNode.initializer)
          : (stylesNode?.initializer ?? $.string('')),
      $styleProps = node.styles?.map((s) =>
        $.createPropertyAssignment($.string(s.name), s.initializer),
      ),
      $varProps = node.vars?.map((s) =>
        $.createPropertyAssignment($.string(`--${s.name}`), s.initializer),
      ),
      styleProps =
        $styleProps || $varProps
          ? runtime.style(base, [...($styleProps ?? []), ...($varProps ?? [])])
          : base;

    props.push($.createPropertyAssignment('style', styleProps));
  } else if (stylesNode) {
    const value = stylesNode.signal
      ? ssrRuntime.unwrap(stylesNode.initializer)
      : stylesNode.initializer;
    props.push($.createPropertyAssignment('style', runtime.style(value, [])));
  }

  if (node.content && node.content.signal) {
    props.push(
      createDangerouslySetInnerHTMLProp(
        node.content.signal
          ? ssrRuntime.unwrap(node.content.initializer)
          : node.content.initializer,
      ),
    );
  }

  return props;
}
