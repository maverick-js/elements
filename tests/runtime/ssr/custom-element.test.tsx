import { CustomElement } from 'maverick.js';
import { defineElement, property } from 'maverick.js/element';
import { renderToString } from 'maverick.js/ssr';

it('should render custom element', () => {
  const element = defineElement({
    tagName: 'mk-foo',
    setup: () => () => <div class="foo">Test</div>,
  });

  const result = renderToString(() => <CustomElement element={element} />).code;

  expect(result).toMatchSnapshot();
});

it('should render custom element with children', () => {
  const element = defineElement({
    tagName: 'mk-foo',
    setup: () => () => <div class="foo">Test</div>,
  });

  const result = renderToString(() => (
    <CustomElement element={element}>
      <div>Child</div>
    </CustomElement>
  )).code;

  expect(result).toMatchSnapshot();
});

it('should render custom element with shadow dom', () => {
  const element = defineElement({
    tagName: 'mk-foo',
    shadow: true,
    setup: () => () => <div class="foo">Test</div>,
  });

  const result = renderToString(() => (
    <CustomElement element={element}>
      <div>Light A</div>
      <div>Light B</div>
      <div>Light C</div>
    </CustomElement>
  )).code;

  expect(result).toMatchSnapshot();
});

it('should render custom element with attributes', () => {
  const element = defineElement({
    tagName: 'mk-foo',
    setup: () => () => null,
  });

  const result = renderToString(() => (
    <CustomElement
      foo={10}
      bar={'boo'}
      $class:foo={true}
      $class:bar={true}
      $style:display={'none'}
      $cssvar:baz={10}
      element={element}
    />
  )).code;

  expect(result).toMatchSnapshot();
});

it('should forward props', () => {
  const element = defineElement({
    tagName: 'mk-foo',
    props: {
      foo: property(10),
    },
    setup:
      ({ props }) =>
      () =>
        <div>{props.foo}</div>,
  });

  const result = renderToString(() => <CustomElement $prop:foo={100} element={element} />).code;

  expect(result).toMatchSnapshot();
});

it('should forward attrs to props', () => {
  const element = defineElement({
    tagName: 'mk-foo',
    props: {
      foo: property(10),
    },
    setup:
      ({ props }) =>
      () =>
        <div>{props.foo}</div>,
  });

  const result = renderToString(() => <CustomElement foo={100} element={element} />).code;

  expect(result).toMatchSnapshot();
});

it('should set inner html', () => {
  const element = defineElement({
    tagName: 'mk-foo',
    setup: () => () => <div>Test</div>,
  });

  const result = renderToString(() => (
    <CustomElement $prop:innerHTML={`<div>INNER HTML</div>`} element={element} />
  )).code;

  expect(result).toMatchSnapshot();
});
