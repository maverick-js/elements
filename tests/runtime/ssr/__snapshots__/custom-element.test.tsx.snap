// Vitest Snapshot v1

exports[`should forward attrs to props 1`] = `"<!$><mk-foo foo=\\"100\\" data-hydrate=\\"\\" data-delegate=\\"\\"><!#internal><div><!$>100</div><!/#internal></mk-foo>"`;

exports[`should forward props 1`] = `"<!$><mk-foo data-hydrate=\\"\\" data-delegate=\\"\\"><!#internal><div><!$>100</div><!/#internal></mk-foo>"`;

exports[`should render custom element 1`] = `"<!$><mk-foo data-hydrate=\\"\\" data-delegate=\\"\\"><!#internal><div class=\\"foo\\">Test</div><!/#internal></mk-foo>"`;

exports[`should render custom element with attributes 1`] = `"<!$><mk-foo foo=\\"10\\" bar=\\"boo\\" data-hydrate=\\"\\" data-delegate=\\"\\" class=\\"foo bar\\" style=\\"display: none;--baz: 10;\\"><!#internal><!/#internal></mk-foo>"`;

exports[`should render custom element with children 1`] = `"<!$><mk-foo data-hydrate=\\"\\" data-delegate=\\"\\"><div>Child</div></mk-foo>"`;

exports[`should render custom element with shadow dom 1`] = `"<!$><mk-foo data-hydrate=\\"\\" data-delegate=\\"\\"><template shadowroot=\\"open\\"><div class=\\"foo\\">Test</div></template><div>Light A</div><div>Light B</div><div>Light C</div></mk-foo>"`;

exports[`should set inner html 1`] = `"<!$><mk-foo data-hydrate=\\"\\" data-delegate=\\"\\"><div>INNER HTML</div></mk-foo>"`;
